src/audio/mod.rs

```diff
@@ -1,44 +1,41 @@
-use crate::{Pipeline, Result};
+use crate::{ring_buffer, Pipeline, Result};
 use anyhow::anyhow;
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
-    Device,
+    Device, StreamConfig,
 };
-use ringbuf::{Consumer, Producer, RingBuffer};
+use ringbuf::RingBuffer;
 use std::time::Duration;
 
 const LATENCY: f32 = 20.0;
 
-pub fn run(mut pipeline: Pipeline) -> Result<()> {
-    let (input_device, output_device) = devices()?;
-
-    let config: cpal::StreamConfig = input_device.default_input_config()?.into();
-
+pub fn run(
+    input_device: &Device,
+    output_device: &Device,
+    config: &StreamConfig,
+    mut pipeline: Pipeline,
```

Comment:
I rearranged where we get the devices and config because we need some of the info from them to create them pedals.